### ISSUE TITLE
How to fine tune your digits_comma file

### ISSUE THREAD

**Issue by carmonasl** [Shreeshrii/tessdata_shreetest#4]
Hello @Shreeshrii  , many thanks for training new files for tesseract, the digits_comma one has improved the accuracy of my model a lot, almost 100% as you can see here https://stackoverflow.com/questions/53866109/pre-processing-image-tesseract-improvement

I'd like to fine tune it with additional fonts so tesseract doesn't mix up 3s, 5s and 9s. Could you please tell me the best procedure to follow for this or a link to a tutorial to know how you did it?

Many thanks again :)

**Comment by safijari** [comment]
This would be very helpful for me as well.

**Comment by Shreeshrii** [comment]
Please see
https://github.com/Shreeshrii/tessdata_shreetest/issues/1#issuecomment-425947492

For details of training done.

Please follow the training wiki page for fine tuning for impact.

I can share the bash script I used, but it is highly dependent on the file
paths.

**Comment by carmonasl** [comment]
@Shreeshrii many thanks for the advice! It'd be awesome if I could use the script as a starting point :)

**Comment by Shreeshrii** [comment]
Please see https://github.com/Shreeshrii/tessdata_shreetest/commit/b69b7e6ba6c7b0bd15f1b5541ac8fa5746383ad4 for the scripts used.

**Comment by Shreeshrii** [comment]
You can also see https://github.com/Shreeshrii/tessdata_ocrb for a sample finetune training with just one font.

**Comment by safijari** [comment]
This is awesome. Thank you so much.

I tend to always get confused in the wiki (maybe I'm not frequenting the right article?) so this would help out a lot.

**Comment by carmonasl** [comment]
Yes thank you very much!!!, @safijari do you mind if we work together to learn faster?

**Comment by carmonasl** [comment]
I get the error "Failed loading language 'eng'". It uses the TESSDATA_PREFIX=/home/alberto/tesseract/tessdata and I can see the eng.traineddata in that folder. Is there anything I need to change? 

I'm sorry for the question @Shreeshrii I'm sure this is quite stupid

#!/bin/bash

~/tesseract/src/training/tesstrain.sh \
  --fonts_dir ~/fonts \
  --lang eng --linedata_only \
  --noextract_font_properties \
  --langdata_dir ~/langdata \
  --tessdata_dir ~/tesseract/tessdata \
  --training_text ./eng.digits.training_text \
  --workspace_dir ~/tmp \
  --output_dir ~/tesstutorial/digits \
   --fontlist  \
    "Abel Regular" \
    "Montserrat Regular" \
    "Roboto Medium"

rm -rf ~/tesstutorial/digits_from_full

**Comment by Shreeshrii** [comment]
Change tessdata_dir ~/tesseract/tessdata
To the directory matching your tessdata_prefix which has eng.traineddata

**Comment by carmonasl** [comment]
I've tried with the OCRB example and I obtain the same result:

23:version:size=80, offset=23466570
Loaded file /home/alberto/tesstutorial/ocrb_from_full/eng.lstm, unpacking...
Warning: LSTMTrainer deserialized an LSTMRecognizer!
Code range changed from 111 to 14!
Num (Extended) outputs,weights in Series:
  1,36,0,1:1, 0
Num (Extended) outputs,weights in Series:
  C3,3:9, 0
  Ft16:16, 160
Total weights = 160
  [C3,3Ft16]:16, 160
  Mp3,3:16, 0
  Lfys64:64, 20736
  Lfx96:96, 61824
  Lrx96:96, 74112
  Lfx512:512, 1247232
  Fc14:14, 0
Total weights = 1404064
Previous null char=110 mapped to 13
Continuing from /home/alberto/tesstutorial/ocrb_from_full/eng.lstm
Loaded 404/404 pages (1-404) of document /home/alberto/tesstutorial/ocrb/eng.Abel_Regular.exp0.lstmf
Iteration 0: ALIGNED TRUTH : 9902013313
Iteration 0: BEST OCR TEXT : . 64 -4 -4 - . . 4 4 . 4 
File /tmp/eng-2018-12-21.lpQ/eng.Abel_Regular.exp0.lstmf page 34 :
!int_mode_:Error:Assert failed:in file ../../../src/lstm/weightmatrix.cpp, line 268
./finetune-ocrb.sh: line 30: 23761 Segmentation fault      (core dumped) OMP_THREAD_LIMIT=1 lstmtraining --model_output ~/tesstutorial/ocrb_from_full/ocrb_plus --traineddata ~/tesstutorial/ocrb/eng/eng.traineddata --continue_from ~/tesstutorial/ocrb_from_full/eng.lstm --old_traineddata ~/tesseract/tessdata/eng.traineddata --train_listfile ~/tesstutorial/ocrb/eng.training_files.txt --debug_interval -1 --max_iterations 410
Failed to read continue from: /home/alberto/tesstutorial/ocrb_from_full/ocrb_plus_checkpoint
/home/alberto/tesstutorial/ocrb_from_full/ocrb_plus_checkpoint is not a recognition model, trying training checkpoint...
Failed to load model from: /home/alberto/tesstutorial/ocrb_from_full/ocrb_plus_checkpoint
Loaded 404/404 pages (1-404) of document /home/alberto/tesstutorial/ocrb/eng.Abel_Regular.exp0.lstmf
Warning: LSTMTrainer deserialized an LSTMRecognizer!
At iteration 0, stage 0, Eval Char error rate=5.3387429, Word error rate=19.315182
Loaded 404/404 pages (1-404) of document /home/alberto/tesstutorial/ocrb/eng.Abel_Regular.exp0.lstmf
Warning: LSTMTrainer deserialized an LSTMRecognizer!
At iteration 0, stage 0, Eval Char error rate=3.4334789, Word error rate=10.338284

**Comment by Shreeshrii** [comment]
You need to use the eng.traineddata from tessdata_best repository for the
extraction of lstm model i.e. to extract /home/alberto/tesstutorial/
ocrb_from_full/eng.lstm

**Comment by carmonasl** [comment]
When I run the makedata script with the original eng.traineddata it works. I've run the second script with the eng.traineddata from tessdata_best and it has generated my new traineddata!

Is there anything wrong with using one file for each script?

**Comment by carmonasl** [comment]
You are awesome man thank you very much!

**Comment by lamchun1110** [comment]
Hi, I'm new to Tesseract. 
Do you have any ideas to make a trainedata with digits and slash?

**Comment by Shreeshrii** [comment]
@lamchun1110 

I have uploaded 2 new traineddata files which have digits, slash as well as some other punctuation too.

see
digits_layer.traineddata	 - replace layer for digits with various punctuation marks
digitsall_layer.traineddata - replace top layer - digits with punctuation -more fonts
